File: frontend/src/types/explore.ts
export interface ExploreRoot {
  id: string;
  label: string;
}

export interface ExploreNode {
  id: string;
  label: string;
  domain: string;
  description: string;
  connection: string;
  surprise_score: number;
}

export interface MemoryProfile {
  anonymous_user_id: string;
  preferred_domains: string[];
  disliked_domains: string[];
  liked_concepts: string[];
  disliked_concepts: string[];
  preferred_surprise_level: number;
  evidence_count: number;
  updated_at: string;
  memory_signature: string;
}

export interface MemoryProfileResponse {
  available: boolean;
  profile: MemoryProfile | null;
}

export interface AgentMetrics {
  memory_retrieval_ms: number;
  llm_latency_ms: number;
  total_latency_ms: number;
  candidate_cache_hit: boolean;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  total_tokens: number | null;
  memory_context_chars: number;
  changed_nodes_vs_baseline: number;
  exploration_ratio: number;
}

export interface ExploreResponseMetadata {
  memory: MemoryProfileResponse | null;
  agent_metrics: AgentMetrics | null;
}

export interface ExploreResponse {
  root: ExploreRoot;
  nodes: ExploreNode[];
  generation_source: string | null;
  metadata: ExploreResponseMetadata | null;
}

export interface ExploreRequest {
  topic: string;
  surprise_level: number;
  anonymous_user_id?: string | null;
  use_memory?: boolean;
}

File: backend/app/models/explore.py
"""Request and response schemas for the explore endpoint."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.memory import AgentMetrics, MemoryProfileResponse


class ExploreRequest(BaseModel):
    """A topic and the user's preferred mock surprise level."""

    model_config = ConfigDict(extra="forbid")

    topic: str = Field(min_length=1, max_length=120)
    surprise_level: float = Field(default=0.5, ge=0, le=1)
    anonymous_user_id: str | None = Field(default=None, max_length=64)
    use_memory: bool = Field(default=True)

    @field_validator("topic", mode="before")
    @classmethod
    def normalize_topic(cls, value: object) -> object:
        """Trim surrounding whitespace before length validation."""

        return value.strip() if isinstance(value, str) else value


class ExploreRoot(BaseModel):
    """The normalized starting point for an exploration."""

    id: str
    label: str


class ExploreNode(BaseModel):
    """A single cross-domain mock recommendation."""

    id: str
    label: str
    domain: str
    description: str
    connection: str
    surprise_score: float = Field(ge=0, le=1)


class ExploreResponseMetadata(BaseModel):
    """Extended metadata for explore responses."""

    model_config = ConfigDict(extra="forbid")

    memory: MemoryProfileResponse | None = None
    agent_metrics: AgentMetrics | None = None


class ExploreResponse(BaseModel):
    """The root topic and its candidate exploration nodes."""

    root: ExploreRoot
    nodes: list[ExploreNode]
    generation_source: str | None = None
    metadata: ExploreResponseMetadata | None = None
